electrumsv/wallet.py

```diff
@@ -5034,6 +5040,42 @@ async def _consume_tip_filter_matches_async(self, state: ServerConnectionState)
             * If the batch is processed then it is no longer valid.
             * If the batch is processed (task cancelled) then it will be picked up next restart.
         """
+        async def add_missing_transactions_from_db() -> None:
+            # We have to go to the database to find out:
+            # - What account a push data match is associated with.
+            # - Which matches do not already have the associated transaction imported.
+            # We could do this in the `create_pushdata_match_async` call and return it, but
+            # this double-dipping in the database is good enough for now.
+            rows_by_account_id: dict[int, list[PushDataMatchMetadataRow]] = {}
+            match_metadata_rows = self.data.read_pushdata_match_metadata()
+            for metadata_row in match_metadata_rows:
+                if metadata_row.account_id in rows_by_account_id:
+                    rows_by_account_id[metadata_row.account_id].append(metadata_row)
+                else:
+                    rows_by_account_id[metadata_row.account_id] = [metadata_row]
+
+            self.logger.debug("Wallet processing %d tip filter matches", len(match_metadata_rows))
+
+            for account_id, match_metadata_rows in rows_by_account_id.items():
+                obtain_transaction_keys: list[MissingTransactionMetadata] = []
+                for metadata_row in match_metadata_rows:
+                    obtain_transaction_keys.append(
+                        MissingTransactionMetadata(metadata_row.transaction_hash, {
+                            TransactionKeyUsageMetadata(metadata_row.pushdata_hash,
+                                metadata_row.keyinstance_id, metadata_row.script_type)},
+                            metadata_row.block_hash is not None))
+                self.logger.debug("Obtaining %d transactions for account %d, %s",
+                    len(obtain_transaction_keys), account_id, obtain_transaction_keys)
+                await self.obtain_transactions_async(account_id, obtain_transaction_keys,
+                    TransactionImportFlag.TIP_FILTER_MATCH)
+
+        # Run a preliminary check to ensure all pushdata notifications in the database
+        # have resulted in a successful fetch of the missing transaction. It's theoretically
+        # possible for an untimely crash to interrupt the `_obtain_transactions_worker_async` task.
+        # In that case, this step will re-add the transaction to the missing transaction cache
+        # at startup to re-attempt the transaction fetch & import.
+        await add_missing_transactions_from_db()
```

Comment:
Why do we care about going to extra lengths to recover from untimely crashes elsewhere? We assert things proactively to make things error if they are broken.

Reply:
It doesn't even need to be a "crash". If the reference server is unavailable, then the transaction will fail to be fetched and if they wallet is then shutdown, there is no other mechanism to re-populate the missing_transactions cache...

Reply:
To be clear, if there is a problem in `_obtain_transactions_worker_async` then it can deal with it. We should not be referring to undescribed problems elsewhere and working around them.

Reply:
I suggest you make a task in the payd backlog describing what the problem is and what you think should be done to fix it. It can be done separately.